finality-aleph/src/data_io.rs

```diff
@@ -359,82 +1001,228 @@ where
+                // The hashes in `branch` are ordered from top to bottom -- need to reverse.
+                branch.reverse();
+                *self.data_to_propose.lock() =
+                    AlephData::HeadProposal(AlephProposal::<B>::new(branch, num_last));
+            } else {
+                // By backtracking from the best block we reached a block conflicting with best finalized.
+                // This is most likely a bug, or some extremely unlikely synchronization issue of the client.
+                warn!(target: "afa", "Error updating data. Conflicting blocks: {:?}, finalized {:?}", curr_block, finalized_block);
+                // We leave the old data intact -- it is not up-to-date, but at least it is correct.
+            }
+        } else {
+            // We leave the default `data_to_propose` -- we don't have the necessary input to compute a new one.
+        }
+    }
+
+    pub(crate) async fn get_best_header(&self) -> B::Header {
+        self.select_chain.best_chain().await.expect("No best chain")
+    }
+
+    pub(crate) async fn run(mut self, mut exit: oneshot::Receiver<()>) {
+        // We would like `best_block_in_session` to correspond to the highest ancestor of the `best_block` (this is what
+        // `select_chain` provides us with) up to the maximal height of `max_block_num`. This task periodically
+        // queries `select_chain` for the `best_block` and updates `best_block_in_session` accordingly. One optimization that it
+        // uses is that once the block in `best_block_in_session` reaches the height of `max_block_num`, and the just queried
+        // `best_block` is a `descendant` of the previous query, then we don't need to update `best_block_in_session`, as it is
+        // already correct. After updating `best_block_in_session`, we call `update_data` to update `data_to_propose`.
+        let mut prev_best_block: Option<BlockHashNum<B>> = None;
+        loop {
+            let delay = futures_timer::Delay::new(Duration::from_millis(REFRESH_INTERVAL));
+            tokio::select! {
+                _ = delay => {
+                    let new_best_header = self.get_best_header().await;
+                    let new_best_block = BlockHashNum::<B>::new(new_best_header.hash(), *new_best_header.number());
+                    if new_best_header.number() <= &self.max_block_num {
+                        self.best_block_in_session = Some(new_best_block.clone());
+                    } else if prev_best_block.is_none() || prev_best_block.as_ref().unwrap().num < self.max_block_num {
+                            // This is either the first iteration in the loop, or the case when the `new_best_header` first
+                            // exceeds `max_block_num`. We don't use any tricks just reduce the `new_best_header` to `max_block_num`
+                            let reduced_header = reduce_header_to_num(self.client.clone(), new_best_header.clone(), self.max_block_num);
```

Comment:
That's correct, but this would actually complicate the logic in this loop and I would rather avoid that. Generally the code of this component is bad (but at least it is battle-tested and it works), and I don't think it's worth investing time in micro-optimizations, one should rather ditch it and write anew, but that is out of scope of my task.